// edge-lexer/build/src/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unclosedTag = exports.unclosedCurlyBrace = exports.unopenedParen = exports.unclosedParen = exports.cannotSeekStatement = void 0;
const edge_error_1 = require("edge-error");
function cannotSeekStatement(chars, pos, filename) {
    return new edge_error_1.EdgeError(`Unexpected token "${chars}"`, 'E_CANNOT_SEEK_STATEMENT', {
        line: pos.line,
        col: pos.col,
        filename: filename,
    });
}
exports.cannotSeekStatement = cannotSeekStatement;
function unclosedParen(pos, filename) {
    return new edge_error_1.EdgeError('Missing token ")"', 'E_UNCLOSED_PAREN', {
        line: pos.line,
        col: pos.col,
        filename: filename,
    });
}
exports.unclosedParen = unclosedParen;
function unopenedParen(pos, filename) {
    return new edge_error_1.EdgeError('Missing token "("', 'E_UNOPENED_PAREN', {
        line: pos.line,
        col: pos.col,
        filename: filename,
    });
}
exports.unopenedParen = unopenedParen;
function unclosedCurlyBrace(pos, filename) {
    return new edge_error_1.EdgeError('Missing token "}"', 'E_UNCLOSED_CURLY_BRACE', {
        line: pos.line,
        col: pos.col,
        filename: filename,
    });
}
exports.unclosedCurlyBrace = unclosedCurlyBrace;
function unclosedTag(tag, pos, filename) {
    return new edge_error_1.EdgeError(`Unclosed tag ${tag}`, 'E_UNCLOSED_TAG', {
        line: pos.line,
        col: pos.col,
        filename: filename,
    });
}
exports.unclosedTag = unclosedTag;
