// @poppinss/file-generator/build/index.js
"use strict";
/*
 * @poppinss/file-generator
 *
 * (c) Poppinss
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateXlsx = exports.generatePng = exports.generatePdf = exports.generateJpg = exports.generateGif = exports.generateDocx = void 0;
var docx_1 = require("./src/files/docx");
Object.defineProperty(exports, "generateDocx", { enumerable: true, get: function () { return docx_1.generateDocx; } });
var gif_1 = require("./src/files/gif");
Object.defineProperty(exports, "generateGif", { enumerable: true, get: function () { return gif_1.generateGif; } });
var jpg_1 = require("./src/files/jpg");
Object.defineProperty(exports, "generateJpg", { enumerable: true, get: function () { return jpg_1.generateJpg; } });
var pdf_1 = require("./src/files/pdf");
Object.defineProperty(exports, "generatePdf", { enumerable: true, get: function () { return pdf_1.generatePdf; } });
var png_1 = require("./src/files/png");
Object.defineProperty(exports, "generatePng", { enumerable: true, get: function () { return png_1.generatePng; } });
var xlsx_1 = require("./src/files/xlsx");
Object.defineProperty(exports, "generateXlsx", { enumerable: true, get: function () { return xlsx_1.generateXlsx; } });
